### PR TITLE
Remove redundant allocation in signatures

### PR DESCRIPTION
**Pull request by l4l**
can be related to #223

~~work still in progress, plz don't merge~~

### CODE REVIEW

File: core/infra/crypto/signature.cpp [hunk 1]
```diff
@@ -27,97 +27,89 @@ limitations under the License.
...
-    std::unique_ptr<T[]> res(new T[sizeof(T)*vec.size()+1]);
-    size_t pos = 0;
-    auto* v = res.get();
-    for(auto c : vec){
-      v[pos++] = c;
-    }
-    v[pos] = '\0';
+template<typename T>
+std::vector<T> vector2UnsignedCharPointer(const std::vector<T> &vec) {
+  std::vector<T> res(vec.size() + 1);
```
Comment (Warchant) [inline core/infra/crypto/signature.cpp]: simple
```
std::vector<T> res{vec.begin(), vec.end()}; 
res.push_back('\0');
```
does not work?

File: core/infra/crypto/signature.cpp [hunk 2]
```diff
@@ -27,97 +27,89 @@ limitations under the License.
...
+                 const KeyPair &keyPair) {
+  std::vector<unsigned char> pub(keyPair.publicKey.begin(), keyPair.publicKey.end());
+  std::vector<unsigned char> pri(keyPair.privateKey.begin(), keyPair.privateKey.end());
+  return base64::encode(sign(message, pub, pri));
+}
+
+std::string sign(const std::string &message,
+                 const std::string &publicKey_b64,
+                 const std::string &privateKey_b64) {
+  auto pub_decoded = base64::decode(publicKey_b64);
```
Comment (Warchant) [inline core/infra/crypto/signature.cpp]: As I remember decode returns vector of unsigned chars? So why do you do let copying?

File: core/infra/crypto/signature.cpp [hunk 3]
```diff
@@ -27,97 +27,89 @@ limitations under the License.
...
+
+  ed25519_create_seed(seed.get());
+  ed25519_create_keypair(
+    publicKeyRaw.get(),
+    privateKeyRaw.get(),
+    seed.get()
+  );
+
+  return KeyPair(
+     pointer2Vector(std::move(publicKeyRaw), 32),
```
Comment (Warchant) [inline core/infra/crypto/signature.cpp]: You defined consts above, so why you don't use them?

File: core/infra/crypto/signature.cpp (hunk 2 above)
Comment (Warchant) [inline core/infra/crypto/signature.cpp]: Double copying*

File: core/infra/crypto/signature.cpp (hunk 3 above)
Comment (Warchant) [inline core/infra/crypto/signature.cpp]: Also make typedef unsigned char uchar;
It is hard to read long lines

File: core/infra/crypto/signature.cpp (hunk 2 above)
Comment (Warchant) [inline core/infra/crypto/signature.cpp]: auto pub = base64::decode(privateKey_b64)
And then pass to sign

File: core/infra/crypto/signature.cpp (hunk 1 above)
Comment (Warchant) [inline core/infra/crypto/signature.cpp]: What is the purpose of this function?? I don't understand

Comment (l4l) [inline core/infra/crypto/signature.cpp]: redundant, forgot to delete

File: core/infra/crypto/signature.cpp (hunk 2 above)
Comment (l4l) [inline core/infra/crypto/signature.cpp]: nope, it's rvo

File: core/infra/crypto/signature.cpp [hunk 4]
```diff
@@ -27,97 +27,89 @@ limitations under the License.
...
+bool verify(const std::vector<unsigned char> &signature,
+            const std::string &message,
+            const std::vector<unsigned char> &publicKey) {
+  return ed25519_verify(signature.data(),
+                        reinterpret_cast<const unsigned char*>(message.c_str()),
+                        message.size(),
+                        publicKey.data());
+}
+
+KeyPair generateKeyPair() {
```
Comment (Warchant) [inline core/infra/crypto/signature.cpp]:
```
std::vector<uchar> pub(PUB_KEY_SIZE);
std::vector<uchar> pri(PRI_KEY_SIZE);
std::vector<uchar> seed(32);

  ed25519_create_seed(&seed[0]);
  ed25519_create_keypair(
    &pub[0],
    &pri[0],
    seed.data()
  );

  return KeyPair(std::move(pub), std::move(pri));
```

a lot cleaner and without copies/pointers

File: core/crypto/signature.hpp [hunk 5]
```diff
@@ -23,36 +23,38 @@ limitations under the License.
...
-    std::string privateKey_b64
-  );
-
-  bool verify(
-    const std::string &signature_b64,
-    const std::string &message,
-    const std::string &publicKey_b64);
-
-  KeyPair generateKeyPair();
+constexpr size_t PRI_KEY_SIZE = 64;
```
Comment (l4l) [inline core/crypto/signature.hpp]: it's already constexpr

File: core/infra/crypto/signature.cpp [hunk 6]
```diff
@@ -27,97 +27,89 @@ limitations under the License.
...
+}
+
+std::string sign(const std::string &message,
+                 const std::string &publicKey_b64,
+                 const std::string &privateKey_b64) {
+  auto pub_decoded = base64::decode(publicKey_b64);
+  auto pri_decoded = base64::decode(privateKey_b64);
+  std::vector<unsigned char> pub(pub_decoded.begin(), pub_decoded.end());
+  std::vector<unsigned char> pri(pri_decoded.begin(), pri_decoded.end());
+  return base64::encode(sign(message, pub, pri));
```
Comment (l4l) [inline core/infra/crypto/signature.cpp]: need a conversion `std::basic_string<char>` to `std::basic_string<unsigned char>`
`std::string` is the first one

File: core/infra/crypto/signature.cpp [hunk 7]
```diff
@@ -58,58 +43,56 @@ std::string sign(const std::string &message,
...
-    seed.get()
-  );
-
-  return KeyPair(
-     pointer2Vector(std::move(publicKeyRaw), 32),
-     pointer2Vector(std::move(privateKeyRaw), 64)
-   );
+  byte_array_t pub(PUB_KEY_SIZE);
+  byte_array_t pri(PRI_KEY_SIZE);
+  byte_array_t seed(SIG_SIZE);
```
Comment (Warchant) [inline core/infra/crypto/signature.cpp]: it is not `SIG_SIZE` (which is signature size).
It is `seed size`, it never changes and should be `= 32`

**Pull request merged**